cli/js/compiler.ts

```diff
@@ -223,10 +223,32 @@ function getExtension(fileName: string, mediaType: MediaType): ts.Extension {
   }
 }
 
-/** A global cache of module source files that have been loaded. */
-const moduleCache: Map<string, SourceFile> = new Map();
-/** A map of maps which cache source files for quicker modules resolution. */
-const specifierCache: Map<string, Map<string, SourceFile>> = new Map();
+/** Because we support providing types for JS files as well as X-TypeScript-Types
+ * header we might be feeding TS compiler with different files than import specifiers
+ * suggest. To accomplish that we keep track of two different specifiers:
+ *  - original - the one in import statement (import "./foo.js")
+ *  - mapped - if there's no type directive it's the same as original, otherwise
+ *             it's unresolved specifier for type directive (/// @deno-types="./foo.d.ts")
+ */
+interface SourceFileSpecifierMap {
+  original: string;
+  mapped: string;
+}
+
+/** A global cache of module source files that have been loaded.
+ * This cache will be rewritten to be populated on compiler startup
+ * with files provided from Rust in request message.
+ */
+const SOURCE_FILE_CACHE: Map<string, SourceFile> = new Map();
+/** A map of maps which cache resolved specifier for each import in a file.
+ * This cache is used so `resolveModuleNames` ops is called as few times
+ * as possible.
+ *
+ * First map's key is "referrer" URL ("file://a/b/c/mod.ts")
+ * Second map's key is "raw" import specifier ("./foo.ts")
+ * Second map's value is resolved import URL ("file:///a/b/c/foo.ts")
+ */
+const RESOLVED_SPECIFIER_CACHE: Map<string, Map<string, string>> = new Map();
```

Comment:
Maybe can be removed. Op call is very cheap relative to tsc

Reply:
It's actually used for handling of type directives